### DISCUSSION TITLE
Generalized ICP feature

### DISCUSSION THREAD

**Discussion by arjay55** [introlab/rtabmap#1149] [Q&A]
Hi,

Is there a Generalized ICP feature on RTABMAP? If there isn't, any advice on which section of the code to add this feature? PCL has a class for this. If it's not too hard, I will fork this and attempt to add this feature, or have this as feature request. Thanks. I am working on my research regarding high accuracy SLAM.

**Comment by matlabbe** [comment]
RTAB-Map implemented three ICP approaches, PCL [default one](http://pointclouds.org/documentation/classpcl_1_1_iterative_closest_point.html), [libpointmatcher](https://libpointmatcher.readthedocs.io/en/latest/) and [CCCorelib](https://github.com/CloudCompare/CCCoreLib) (the one used in [CloudCompare](https://www.danielgm.net/cc/)). We can chose which one we want by using parameter `Icp/Strategy` parameter. 
```
Param: Icp/Strategy = "1"  [ICP implementation: 0=Point Cloud Library, 1=libpointmatcher, 2=CCCoreLib (CloudCompare).]
```

For Generalized ICP, do you mean https://pointclouds.org/documentation/classpcl_1_1_generalized_iterative_closest_point.html ?

For `Icp/Strategy=0`, we have wrapping functions for PointToPoint and PointToPLane approaches here:
https://github.com/introlab/rtabmap/blob/71bb0cf2262ed73c937312da3b3c7cc8d0ad0f13/corelib/src/util3d_registration.cpp#L386-L492

These functions are used in [RegistrationICP](https://github.com/introlab/rtabmap/blob/master/corelib/src/RegistrationIcp.cpp) when `Icp/Strategy=0`. You could easily wrap a new function called `util3d::gicp(...)` using `pcl::gicp(...)`, then check where `util3d::icp` and `util3d::icpPointToPlane` are called and add new `if(strategy_ == 4) { // do GICP} else {//do current strategy=0 functions...}`. If you need to expose GICP parameters in RTAB-Map, look how we prefixed specific Icp parameters:
https://github.com/introlab/rtabmap/blob/71bb0cf2262ed73c937312da3b3c7cc8d0ad0f13/corelib/include/rtabmap/core/Parameters.h#L714-L757
